tools/codegen/model.py

```diff
@@ -112,6 +112,19 @@ class NativeFunction:
     # defined.  This is for conveniently reporting error messages!
     loc: 'Location'
 
+    # Whether or not this out functions is a "structured kernel".  Structured
+    # kernels are defined a little differently from normal kernels; in
+    # particular, their shape checking logic is defined separately from
+    # the kernel.  Only out functions can be structured; other functions
+    # delegate to the out function using the structured_delegate keyword.
+    # Every structured kernel must have at least an out and a functional
+    # variant.
+    structured: bool
+
+    # Whether or not this non-out function is a structured kernel, defined
+    # in terms of the out kernel referenced by the string here.
+    structured_delegate: Optional['OperatorName']
```

Comment:
Data modeling OCD obliges me to note the room for nonsense in having both these properties 😁 

In a perfect world we'd have a single property whose type depended on `func.kind`, but in the absence of that, there's no great choice (assuming making `SchemaKind` into an ADT that carries this information directly is either impossible in mypy, or unwanted for some other reason).

Half-hearted pitch for 
```
structured: Optional['OperatorName']
```
and out functions just use `''` or whatever and `is_structured()` just tests `is not None`. `get_structured_delegate()` could do the obvious.

Reply:
Yeah, this choice was driven by what I wanted surface syntax to look like, and then trying to keep NativeFunction class looking similar. What do you think is clearer:

```
- func: foo.out
  structured: None

- func: foo.Tensor
  structured: foo.out
```

or 

```
- func: foo.out
  structured: true

- func: foo.Tensor
  structured_delegate: foo.out
```

Reply:
#2. I think "`structured: None` means it's structured" would be hella confusing. (More generally I think it's better in the surface syntax to have the attributes be, like, monomorphic :)

Matching representation to surface syntax is a nice through line - might be worth a one-liner in the comment header for `NativeFunction` (apologies if it's there already and I missed it)